File: app/src/test/java/kztproject/jp/splacounter/repository/AuthRepositoryTest.java
package kztproject.jp.splacounter.repository;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;

import io.reactivex.Single;
import kztproject.jp.splacounter.DummyCreator;
import kztproject.jp.splacounter.api.MyServiceClient;
import kztproject.jp.splacounter.model.UserResponse;
import kztproject.jp.splacounter.preference.AppPrefs;
import kztproject.jp.splacounter.preference.AppPrefsProvider;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class AuthRepositoryTest {

    private MyServiceClient mockClient = mock(MyServiceClient.class);

    private AppPrefsProvider mockAppPrefsProvider = new AppPrefsProvider(RuntimeEnvironment.application);

    private AuthRepository repository;

    @Before
    public void setup() {
        repository = new AuthRepository(mockClient, mockAppPrefsProvider);
    }

    @Test
    public void login() {
        UserResponse dummyResponse = DummyCreator.createDummyUserResponse();
        when(mockClient.getUser(anyString()))
                .thenReturn(Single.just(dummyResponse));

        repository.login("test")
                .test()
                .assertNoErrors()
                .assertComplete();

        AppPrefs appPrefs = AppPrefs.get(RuntimeEnvironment.application);
        assertThat(appPrefs.getUserId(), is(dummyResponse.user.id));
        assertThat(appPrefs.getUserName(),is(dummyResponse.user.fullName));
    }
}

File: app/src/test/java/kztproject/jp/splacounter/repository/AuthRepositoryTest.kt
package kztproject.jp.splacounter.repository

import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.Single
import kztproject.jp.splacounter.DummyCreator
import kztproject.jp.splacounter.api.RewardListClient
import kztproject.jp.splacounter.auth.api.TodoistService
import kztproject.jp.splacounter.auth.repository.AuthRepository
import kztproject.jp.splacounter.model.RewardUser
import kztproject.jp.splacounter.model.UserResponse
import kztproject.jp.splacounter.preference.PrefsWrapper
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyLong
import org.mockito.Mockito.anyString
import org.robolectric.RobolectricTestRunner
import org.robolectric.RuntimeEnvironment

@RunWith(RobolectricTestRunner::class)
class AuthRepositoryTest {

    private val mockClient = mock<TodoistService>()
    private val mockRewardListClient = mock<RewardListClient>()

    private lateinit var repository: AuthRepository

    @Before
    fun setup() {
        PrefsWrapper.initialize(RuntimeEnvironment.application)
        repository = AuthRepository(mockClient, mockRewardListClient)
    }

    @Test
    fun login() {
        val dummyResponse: UserResponse = DummyCreator.createDummyUserResponse()
        val dummyRewardUser: RewardUser = DummyCreator.createDummyRewardUser()
        whenever(mockClient.getUser(anyString(), anyString(), anyString())).thenReturn(Single.just(dummyResponse))
        whenever(mockRewardListClient.findUser(anyLong())).thenReturn(Single.just(dummyRewardUser))

        repository.login("test")
                .test()
                .assertNoErrors()
                .assertComplete()

        assertThat(PrefsWrapper.userId).isEqualTo(dummyRewardUser.id)
    }

    @Test
    fun signUp() {
        val dummyResponse: UserResponse = DummyCreator.createDummyUserResponse()
        val dummyRewardUser: RewardUser = DummyCreator.createDummyRewardUser()
        whenever(mockClient.getUser(anyString(), anyString(), anyString())).thenReturn(Single.just(dummyResponse))
        whenever(mockRewardListClient.createUser(anyLong())).thenReturn(Single.just(dummyRewardUser))

        repository.signUp("test")
                .test()
                .assertNoErrors()
                .assertComplete()

        assertThat(PrefsWrapper.userId).isEqualTo(dummyRewardUser.id)
    }
}